src/plugins/tonegen/plugin.c

```diff
@@ -22,75 +25,186 @@
+    int                 sample_rate;
+    bool                statistics;
+    int                 buflen;
+    int                 minreq;
+    char               *dtmf_tags;
+    char               *ind_tags;
+    int                 dtmf_volume;
+    int                 ind_volume;
+    bool                legacy_api;
+};
+
+struct userdata {
+    NPlugin *plugin;
+    struct properties properties;
+    struct tonegend tonegend;
+};
+
+struct options_parse;
+typedef bool (*prop_value_parser)(const NValue *val, struct options_parse *opt);
+
+struct options_parse {
+    const char *arg_name;
+    prop_value_parser arg_parser;
+    void *arg_value;
+    char **arg_str;
+};
+
+static struct userdata u;
+
+static bool prop_8khz_parser(const NValue *val, struct options_parse *opt);
+static bool prop_standard_parser(const NValue *val, struct options_parse *opt);
+static bool prop_string_parser(const NValue *val, struct options_parse *opt);
+static bool prop_int_parser(const NValue *val, struct options_parse *opt);
+static bool prop_bool_parser(const NValue *val, struct options_parse *opt);
+
+static struct options_parse options[] = {
+    { "8kHz"            , prop_8khz_parser      , &u.properties.sample_rate, NULL },
+    { "standard"        , prop_standard_parser  , &u.properties.standard, NULL    },
+    { "buflen"          , prop_int_parser       , &u.properties.buflen, NULL      },
+    { "minreq"          , prop_int_parser       , &u.properties.minreq, NULL      },
+    { "statistics"      , prop_bool_parser      , &u.properties.statistics, NULL  },
+    { "tag-dtmf"        , prop_string_parser    , NULL, &u.properties.dtmf_tags   },
+    { "tag-indicator"   , prop_string_parser    , NULL, &u.properties.ind_tags    },
+    { "volume-dtmf"     , prop_int_parser       , &u.properties.dtmf_volume, NULL },
+    { "volume-indicator", prop_int_parser       , &u.properties.ind_volume, NULL  },
+    { "legacy-dbus-api" , prop_bool_parser      , &u.properties.legacy_api, NULL  },
+    { NULL              , NULL                  , NULL, NULL                      }
+};
+
+/* OPTION PARSING */
+
+static bool
+prop_8khz_parser (const NValue *val, struct options_parse *opt)
+{
+    struct options_parse b;
+    bool use_8khz;
 
-static DBusConnection *system_bus = NULL;
+    b.arg_value = (void *) &use_8khz;
+
+    if (prop_bool_parser (val, &b)) {
+        if (*(bool *) opt->arg_value)
+            *(int *) opt->arg_value = 8000;
+        return true;
+    } else
+        return false;
+}
 
-static gboolean
-call_dbus_method (DBusConnection *bus, DBusMessage *msg)
+static bool
+prop_bool_parser (const NValue *val, struct options_parse *opt)
 {
-    if (!dbus_connection_send (bus, msg, 0)) {
-        N_WARNING (LOG_CAT "failed to send DBus message %s",
-            dbus_message_get_member (msg), dbus_message_get_interface (msg));
+    const char *str;
+
+    str = n_value_get_string (val);
+    if (g_ascii_strncasecmp (str ?: "", "true", 4) == 0 ||
```

Comment:
This will also interpret "truefoo" as "true", right (due to n=4). Probably okay, though.